store/postgres/src/relational_queries.rs

```diff
@@ -2008,37 +2023,84 @@ impl<'a, Conn> RunQueryDsl<Conn> for FindQuery<'a> {}
 
 #[derive(Debug, Clone)]
 pub struct FindRangeQuery<'a> {
-    table: &'a Table,
-    eb_range: EntityBlockRange,
+    tables: &'a Vec<&'a Table>,
+    causality_region: CausalityRegion,
+    is_upper_range: bool,
+    imm_range: EntityBlockRange,
+    mut_range: EntityBlockRange,
 }
 
 impl<'a> FindRangeQuery<'a> {
-    pub fn new(table: &'a Table, block_range: Range<BlockNumber>) -> Self {
-        let eb_range = EntityBlockRange::new(&table, block_range);
-        Self { table, eb_range }
+    pub fn new(
+        tables: &'a Vec<&Table>,
+        causality_region: CausalityRegion,
+        is_upper_range: bool,
+        block_range: Range<BlockNumber>,
+    ) -> Self {
+        let imm_range = EntityBlockRange::new(true, block_range.clone(), false);
+        let mut_range = EntityBlockRange::new(false, block_range, is_upper_range);
+        Self {
+            tables,
+            causality_region,
+            is_upper_range,
+            imm_range,
+            mut_range,
+        }
     }
 }
 
 impl<'a> QueryFragment<Pg> for FindRangeQuery<'a> {
     fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, Pg>) -> QueryResult<()> {
         out.unsafe_to_cache_prepared();
+        let mut first = true;
 
-        // Generate
-        //    select '..' as entity, to_jsonb(e.*) as data
-        //      from schema.table e where id = $1
-        out.push_sql("select ");
-        out.push_bind_param::<Text, _>(self.table.object.as_str())?;
-        out.push_sql(" as entity, to_jsonb(e.*) as data\n");
-        out.push_sql("  from ");
-        out.push_sql(self.table.qualified_name.as_str());
-        out.push_sql(" e\n where ");
-        // TODO: do we need to care about it?
-        // if self.table.has_causality_region {
-        //     out.push_sql("causality_region = ");
-        //     out.push_bind_param::<Integer, _>(&self.key.causality_region)?;
-        //     out.push_sql(" and ");
-        // }
-        self.eb_range.contains(&mut out)
+        for table in self.tables.iter() {
+            // the immutable entities don't have upper range and also can't be modified or deleted
+            if !(self.is_upper_range && table.immutable) {
+                if first {
+                    first = false;
+                } else {
+                    out.push_sql("\nunion all\n");
+                }
+
+                // Generate
+                //    select '..' as entity, to_jsonb(e.*) as data, block$ as block_number
```

Comment:
For mutable tables, what is `block_number` set to? (Looking at the code below, I think I can work it out, but would be nice to have it in the comment)